types/src/network_address.rs

```diff
@@ -0,0 +1,645 @@
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        #[derive(Serialize)]
+        #[serde(rename = "NetworkAddress")]
+        struct SerializeWrapper<'a>(&'a [Protocol]);
+
+        if serializer.is_human_readable() {
+            serializer.serialize_str(&self.to_string())
+        } else {
+            let val = SerializeWrapper(&self.0.as_ref());
+            val.serialize(serializer)
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for NetworkAddress {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct StringVisitor;
+
+        impl<'de> de::Visitor<'de> for StringVisitor {
+            type Value = NetworkAddress;
+
+            fn expecting(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                write!(f, "NetworkAddress")
+            }
+
+            fn visit_str<E: de::Error>(self, v: &str) -> Result<Self::Value, E> {
+                NetworkAddress::from_str(v).map_err(de::Error::custom)
+            }
+        }
+
+        #[derive(Deserialize)]
+        #[serde(rename = "NetworkAddress")]
+        struct DeserializeWrapper(Vec<Protocol>);
+
+        if deserializer.is_human_readable() {
+            deserializer.deserialize_str(StringVisitor)
+        } else {
+            let wrapper = DeserializeWrapper::deserialize(deserializer)?;
+            let addr = NetworkAddress::try_from(wrapper.0).map_err(de::Error::custom)?;
+            Ok(addr)
+        }
+    }
+}
+
+#[cfg(any(test, feature = "fuzzing"))]
+impl Arbitrary for NetworkAddress {
+    type Parameters = ();
+    type Strategy = BoxedStrategy<Self>;
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        vec(any::<Protocol>(), 1..10)
+            .prop_map(NetworkAddress::new)
+            .boxed()
+    }
+}
+
+//////////////
+// Protocol //
+//////////////
+
+impl fmt::Display for Protocol {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use self::Protocol::*;
+        match self {
+            Ip4(addr) => write!(f, "/ip4/{}", addr),
+            Ip6(addr) => write!(f, "/ip6/{}", addr),
+            Dns(domain) => write!(f, "/dns/{}", domain),
+            Dns4(domain) => write!(f, "/dns4/{}", domain),
+            Dns6(domain) => write!(f, "/dns6/{}", domain),
+            Tcp(port) => write!(f, "/tcp/{}", port),
+            Memory(port) => write!(f, "/memory/{}", port),
+            NoiseIk(pubkey) => write!(
+                f,
+                "/ln-noise-ik/{}",
+                pubkey
+                    .to_encoded_string()
+                    .expect("ValidKeyStringExt::to_encoded_string is infallible")
+            ),
+            Handshake(version) => write!(f, "/ln-handshake/{}", version),
+        }
+    }
+}
+
+/////////////
+// DnsName //
+/////////////
+
+impl DnsName {
+    pub fn is_valid(s: &str) -> Result<(), ParseError> {
```

Comment:
Not a big deal but I usually think `bool` when i see a function named `is_*`

Reply:
ahh you're right, that's a bit confusing